installinator/src/reporter.rs

```diff
@@ -49,129 +78,418 @@ impl ProgressReporter {
+            let (discover_cancel_tx, discover_cancel_rx) =
+                coop_cancel::new_pair();
+            let (report_tx, report_rx) = watch::channel(None);
+            let (completion_tx, mut completion_rx) = mpsc::channel(1);
+            self.discover_task =
+                Some(self.spawn_discover_task(peers_tx, discover_cancel_rx));
+            self.reconcile_peers_task =
+                Some(self.spawn_reconcile_peers_task(peers_rx, report_rx));
+
+            let mut reports_done = false;
             let mut events_done = false;
 
+            // The goal of this loop is to:
+            //
+            // * process all events
+            // * get a fully completed report out to at least one peer which
+            //   accepts this report
+            //
+            // The state machine is implemented via a `tokio::select` loop. The
+            // general logic is:
+            //
+            // * on receiving an event from the update engine
+            //   - add event to self.buffer
+            //   - if event receiver is closed, the update engine has completed, so
+            //     mark events_done = true, and indicate that the next report should
+            //     be sent right away
+            //
+            // * on report_interval tick:
+            //   - generate report
+            //   - if events_done is true, mark report as complete and pass in a
+            //     completion channel sender
+            //   - send message to report tasks
+            //
+            // * on receiving completion channel message (indicating that at
+            //   least one completed report has been received by a peer):
+            //   - mark reports_done = true
+            //
+            // If events_done and reports_done are true, then the loop's goal is
+            // met, so exit the loop.
             loop {
                 tokio::select! {
-                    _ = interval.tick() => {
-                        self.on_tick().await;
-                    }
-
                     event = self.event_receiver.recv(), if !events_done => {
                         if let Some(event) = event {
                             self.buffer.add_event(event);
                         } else {
-                            // The completion sender has been dropped; this is
-                            // an indication to drain the completion queue and
-                            // end the loop.
+                            // The event sender has been dropped; this is an
+                            // indication to send one final report. In order to
+                            // do so, reset the report interval to now so that
+                            // the next loop iteration can send the report right
+                            // away.
                             events_done = true;
+                            report_interval.reset_at(Instant::now());
```

Comment:
Maybe every so slightly more clear as

```suggestion
                            report_interval.reset_immediately();
```

Reply:
oh neat, didn't know about that -- thanks!